js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,412 @@
+  interest.setChildSelector(0);
+  interest.setMustBeFresh(false);
+
+  while (this.nextSegmentToRequest <= this.finalBlockNo && this.segmentsOnFly <= this.windowSize) {
+    // Start with the original Interest to preserve any special selectors.
+    interest.setName(dataName.getPrefix(-1).appendSegment(this.nextSegmentToRequest));
+    interest.refreshNonce();
+
+    var thisPipeline = this;
+
+    this.fetchSegment(interest);
+
+    this.nextSegmentToRequest += 1;
+    this.increaseNumberOfSegmentsOnFly;
+  }
+  this.sendInterestDelay += ((Date.now() - sTime)/1000);
+};
+
+PipelineFixed.prototype.onData = function(originalInterest, data)
+{
+  var sTime = Date.now();
+
+  if (this.validatorKeyChain != null) {
+    try {
+      var thisPipeline = this;
+      this.validatorKeyChain.verifyData
+        (data,
+         function(localData) {
+           thisPipeline.onVerified(localData, originalInterest);
+         },
+         this.onValidationFailed.bind(this));
+    } catch (ex) {
+      console.log("Error in KeyChain.verifyData: " + ex);
+    }
+  }
+  else {
+    if (!this.verifySegment(data)) {
+      try {
+        this.onError
+          (PipelineFixed.ErrorCode.SEGMENT_VERIFICATION_FAILED,
+           "Segment verification failed");
+      } catch (ex) {
+        console.log("Error in onError: " + NdnCommon.getErrorWithStackTrace(ex));
+      }
+      return;
+    }
+
+    this.onDataDelay += ((Date.now() - sTime)/1000);
+    this.onVerified(data, originalInterest);
+  }
+};
+
+PipelineFixed.prototype.onVerified = function(data, originalInterest)
+{
+  var sTime = Date.now();
+  var currentSegment = 0;
+
+  if (!PipelineFixed.endsWithSegmentNumber(data.getName())) {
+    // We don't expect a name without a segment number.  Treat it as a bad packet.
+    try {
+      this.onError
+        (PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+         "Got an unexpected packet without a segment number: " +
+          data.getName().toUri());
+    } catch (ex) {
+      console.log("Error in onError: " + NdnCommon.getErrorWithStackTrace(ex));
+    }
+  }
+  else {
+    try {
+      currentSegment = data.getName().get(-1).toSegment();
+    }
+    catch (ex) {
+      try {
+        this.onError
+          (PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+           "Error decoding the name segment number " +
+           data.getName().get(-1).toEscapedString() + ": " + ex);
+      } catch (ex) {
+        console.log("Error in onError: " + NdnCommon.getErrorWithStackTrace(ex));
+      }
+      return;
+    }
+  }
+
+  // Check for first segment 
+  if (this.firstSegmentIsReceived == false) {
+    this.firstSegmentIsReceived = true;
+    if (currentSegment == 0) {
+      this.nextSegmentToRequest += 1;
+    }
+  }
+ 
+  if (data.getMetaInfo().getFinalBlockId().getValue().size() > 0) {
+    try {
+      this.finalBlockNo = (data.getMetaInfo().getFinalBlockId().toSegment());
+    }
+    catch (ex) {
+      try {
```

Comment:
This “try { onError() } catch { log }” pattern has been repeated several times. Abstract it into a `PipelineFixed.prototype.reportError` function, and remove all duplicates with a call to that function.